noun/retrieve.c

```diff
@@ -1770,113 +1518,122 @@ u3_mur_words(const c3_w* key_w, c3_w len_w)
   }
 }
 
-/* u3_mur_both():
+/* u3r_mug_bytes(): Compute the mug of `buf`, `len`, LSW first.
 **
-**   Join two murs.
+**   XX temporary, reconcile with u3r_mug_words
+*/
+c3_w
+u3r_mug_bytes(const c3_y *buf_y,
+              c3_w        len_w)
+{
+  c3_w syd_w = 0xcafebabe;
+
+  while ( 1 ) {
+    c3_w haz_w, ham_w;
+
+    MurmurHash3_x86_32(buf_y, len_w, syd_w, &haz_w);
+    ham_w = (haz_w >> 31) ^ (haz_w & 0x7fffffff);
+
+    if ( 0 != ham_w ) return ham_w;
+    else syd_w++;
+  }
+}
+
+/* u3r_mug_d(): Compute the mug of `num`, LSW first.
+*/
+c3_w
+u3r_mug_d(c3_d num_d)
+{
+  c3_w buf_w[2];
+
+  buf_w[0] = (c3_w)(num_d & 0xffffffffULL);
+  buf_w[1] = (c3_w)(num_d >> 32ULL);
+
+  return u3r_mug_words(buf_w, 2);
+}
+
+/* u3r_mug_both(): Join two mugs.
 */
 c3_w
-u3_mur_both(c3_w lef_w, c3_w rit_w)
+u3r_mug_both(c3_w lef_w, c3_w rit_w)
 {
   c3_w ham_w = lef_w ^ (0x7fffffff ^ rit_w);
 
-  return u3_mur_words(&ham_w, (0 == ham_w) ? 0 : 1);
+  return u3r_mug_words(&ham_w, (0 == ham_w) ? 0 : 1);
```

Comment:
I don't understand what the 0 case does on this line. It looks like it's taking the mug of a 32-bit 0 with a length of 0 ... what does that mean?

Reply:
Compare raw Murmur3 with a seed and key of 0, but different bytewidths:

```
> (muk 0 0 0)
0
> (muk 0 1 0)
1.364.076.727
> (muk 0 2 0)
821.347.078
```

This is the same thing -- are we passing a zero'd out 32bit buffer, or a buffer of zero length? If we want to support mug calculations over zero'd buffers of some length, we can't move the default into the function. Unfortunately, it to be placed at the `u3r_mug_words` call sites, at least given this function signature.